### ISSUE TITLE
Add support for cloudflare workers environment

### ISSUE THREAD

**Issue by karim23657** [mtcute/mtcute#54]
I found this comment https://github.com/alik0211/mtproto-core/pull/229#issuecomment-1262340871 that has code for storage and transport on cloudflare workers , these are codes : https://github.com/Legend1991/mtproto-core/tree/master/envs/cloudflare

Could you please give me a code example how can i use that (exp. usage in my-awesome-bot template)?

**Comment by teidesu** [comment]
mtcute does not use that library

**Comment by teidesu** [comment]
oops accident

**Comment by teidesu** [comment]
> Error: Cannot perform I/O on behalf of a different request. I/O objects (such as streams, request/response bodies, and others) created in the context of one request handler cannot be accessed from a different request's handler. This is a limitation of Cloudflare Workers which allows us to improve overall performance. (I/O type: Native)

yeah i don't think it is even viable to run mtcute as-is in workers. 
overhead from starting a client on every single request is a bit too much:
<img width="63" alt="image" src="https://github.com/mtcute/mtcute/assets/86301490/b0be4e47-e7b4-4e2d-8c62-e399bf2d4c3f">

there *is* an option to use durable objects (which i will probably eventually get to implementing), but it's not available on free plans so it's probably not much of a priority right now. 

i'll probably eventually implement something like a "remote", where you can connect to a remote mtcute instance in serverless environment, but that requires making better networking abstractions for stuff to work across runtimes. 
i need to do that either way, but rn i want to focus on improving highlevel api coverage and dispatcher dx for a bit

**Comment by SunsetTechuila** [comment]
> there is an option to use durable objects (which i will probably eventually get to implementing), but it's not available on free plans so it's probably not much of a priority right now.

durable objects now available on Workers Free plan: https://developers.cloudflare.com/durable-objects/platform/pricing/

**Comment by SunsetTechuila** [comment]
https://developers.cloudflare.com/changelog/2025-04-07-durable-objects-free-tier/

**Comment by laosb** [comment]
With a bit of messing with WASM loading on Durable Object, I got mtcute working in Durable Object. However, the main problem I think is when to trigger the execution of Durable Object. If I'm understanding the pricing of DO correctly, a $5/mo Workers Paid program can only get one DO running 24/7 for a month, and Durable Object is not really designed for that. 

The immediate thought is to use Bot API webhook to trigger execution of DO, however Telegram officially said that multiple session on bot accounts are not supported and will likely make both sessions miss updates. If your bot doesn't need on-demand triggering by commands / message, it's totally feasible to use cron on workers to trigger the bot on a schedule, but otherwise I think it doesn't make sense.
